File: app/src/main/java/com/remotearthsolutions/expensetracker/utils/CheckoutUtils.java
package com.remotearthsolutions.expensetracker.utils;

import android.app.Activity;
import com.remotearthsolutions.expensetracker.activities.ApplicationObject;
import org.solovyev.android.checkout.*;

public class CheckoutUtils {

    private ActivityCheckout mCheckout;
    private static CheckoutUtils instance;
    private boolean isStarted;

    public static CheckoutUtils getInstance(Activity activity) {
        if (instance == null) {
            instance = new CheckoutUtils(activity);
        }

        return instance;
    }

    public static void clearInstance() {
        instance = null;
    }

    private CheckoutUtils(Activity activity) {
        mCheckout = Checkout.forActivity(activity, ApplicationObject.get().getBilling());
    }

    public void start() {
        if (!isStarted) {
            mCheckout.start();
            isStarted = true;
        }
    }

    public void stop() {
        mCheckout.stop();
        isStarted = false;
    }

    public void createPurchaseFlow(EmptyRequestListener<Purchase> purchaseEmptyRequestListener) {
        mCheckout.createPurchaseFlow(purchaseEmptyRequestListener);
    }

    public PurchaseFlow getPurchaseFlow() {
        return mCheckout.getPurchaseFlow();
    }

    public void load(Inventory.Callback callback, String... productId) {
        Inventory mInventory = mCheckout.makeInventory();
        mInventory.load(Inventory.Request.create()
                .loadAllPurchases()
                .loadSkus(ProductTypes.IN_APP, productId), callback);
    }

    public ActivityCheckout getCheckout() {
        return mCheckout;
    }
}

File: app/src/main/java/com/remotearthsolutions/expensetracker/utils/CheckoutUtils.kt
package com.remotearthsolutions.expensetracker.utils

import android.app.Activity
import android.util.Log
import com.remotearthsolutions.expensetracker.activities.ApplicationObject
import org.solovyev.android.checkout.*

class CheckoutUtils private constructor(activity: Activity) {
    val checkout: ActivityCheckout = Checkout.forActivity(activity, (activity.application as ApplicationObject).billing)
    private var isStarted = false
    private var purchaseFlowCreated = false
    fun start() {
        try {
            if (!isStarted) {
                checkout.start()
                isStarted = true
            }
        } catch (e: Exception) {
            Log.d("Exception", "" + e.message)
        }
    }

    fun stop() {
        checkout.stop()
        checkout.destroyPurchaseFlow()
        purchaseFlowCreated = false
        isStarted = purchaseFlowCreated
    }

    fun createPurchaseFlow(purchaseEmptyRequestListener: EmptyRequestListener<Purchase>?) {
        if (!purchaseFlowCreated) {
            checkout.createPurchaseFlow(purchaseEmptyRequestListener!!)
            purchaseFlowCreated = true
        }
    }

    val purchaseFlow: PurchaseFlow
        get() = checkout.purchaseFlow

    fun load(callback: Inventory.Callback?, vararg productId: String?) {
        val mInventory = checkout.makeInventory()
        mInventory.load(
            Inventory.Request.create()
                .loadAllPurchases()
                .loadSkus(ProductTypes.IN_APP, *productId), callback!!
        )
    }

    companion object {
        private var instance: CheckoutUtils? = null
        fun getInstance(activity: Activity): CheckoutUtils? {
            if (instance == null) {
                instance =
                    CheckoutUtils(activity)
            }
            return instance
        }

        fun clearInstance() {
            instance = null
        }
    }

}